outlines/fsm/fsm.py

```diff
@@ -0,0 +1,216 @@
+    """FSM to generate text that is in the language of a regular expression."""
+
+    def __init__(
+        self,
+        regex_string: str,
+        tokenizer: "Tokenizer",
+        max_tokens: Optional[int] = None,
+    ):
+        regex_pattern = interegular.parse_pattern(regex_string)
+        regex_fsm, _ = make_deterministic_fsm(regex_pattern.to_fsm().reduce())
+        (
+            self.states_to_token_maps,
+            self.empty_token_ids,
+        ) = create_fsm_index_tokenizer(regex_fsm, tokenizer)
+
+        if not any(
+            regex_fsm.finals.intersection(v.values())
+            for v in self.states_to_token_maps.values()
+        ):
+            raise ValueError(
+                "The vocabulary does not allow us to build a sequence that matches the input regex"
+            )
+
+        self.final_states = regex_fsm.finals | {
+            -1
+        }  # Include the EOS token in final states
+        self.max_tokens = max_tokens
+        self.num_tokens_generated = 0
+        self.vocabulary = tokenizer.vocabulary.values()
+        self.end_token = tokenizer.eos_token_id
+
+    def next_instruction(self, state: FSMState) -> List[int]:
+        """Generate a list of forbidden tokens for the next step.
+
+        The initialization of the FSM builds an index which maps FSM states to a
+        map from authorized tokens to the state in which the FSM needs to move
+        if said token is generated. Therefore the authorized tokens at the
+        current state are the keys of the map returned by the value of the index
+        for current state.
+
+        If the current state is not contained in the end this means that we are
+        in a final state of the FSM. We only authorize EOS tokens in the final
+        state.
+
+        Parameters
+        ----------
+        state
+            The current state of the FSM
+
+        Returns
+        -------
+        A list that contains the tokens to mask.
+
+        """
+        next_tokens_to_end_states = self.states_to_token_maps.get(state)
+
+        if next_tokens_to_end_states is None:
+            authorized_tokens = [self.end_token]
+        else:
+            authorized_tokens = list(next_tokens_to_end_states.keys())
+
+        forbidden_tokens = [
+            token for token in self.vocabulary if token not in authorized_tokens
+        ]
+
+        return list(forbidden_tokens)
+
+    def next_state(self, state: FSMState, token_id: int) -> FSMState:
+        """Update the state of the FSM.
+
+        We use the index to determine to which state the FSM should transition
+        given the token that was just generated.
+
+        Parameters
+        ----------
+        state
+            The current state of the FSM.
+        token_id
+            The id of the token that was just generated.
+
+        Returns
+        -------
+        The new state of the FSM.
+
+        """
+        self.num_tokens_generated += 1
+
+        if self.max_tokens is not None:
```

Comment:
I am not sure whether the `max_token` should live here, or closer to the generation. We current can only set the value when initialising the generator, but it may be useful to be able to specify it when *calling* the generator.

Same question regarding stopping once a series of tokens (or characters) has been found.